From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mohamed Ahmed <mohamedahmedegypt2001@gmail.com>
Date: Tue, 25 Aug 2026 22:20:30 +0400
Subject: drm/nouveau/disp: expose armed head state over NVIF

Add NVIF_HEAD_V0_ARMED to return the armed head state to recover firmware
scanout. There was already SCANOUTPOS but it does not have the pixel
clock and viewport.

A nonzero raster alone cannot identify an active head because some VBIOS'
leave stale totals on unused heads. Return -ENODEV when either total or
the pixel clock is zero so callers do not inherit those heads.

Signed-off-by: Mohamed Ahmed <mohamedahmedegypt2001@gmail.com>
(cherry picked from commit a71a5ce492cc01c145872fb18580a160af2064d8)
---
 drivers/gpu/drm/nouveau/include/nvif/head.h   |  3 ++
 drivers/gpu/drm/nouveau/include/nvif/if0013.h | 34 ++++++++++++++
 drivers/gpu/drm/nouveau/nvif/head.c           | 12 +++++
 .../gpu/drm/nouveau/nvkm/engine/disp/uhead.c  | 46 +++++++++++++++++++
 4 files changed, 95 insertions(+)

diff --git a/drivers/gpu/drm/nouveau/include/nvif/head.h b/drivers/gpu/drm/nouveau/include/nvif/head.h
index 3ec36999e9560d2d35f6c6b7b2190be8acbbdbdd..1861ed5dec8448e59b47fb1c42f4be87b9f329cd 100644
--- a/drivers/gpu/drm/nouveau/include/nvif/head.h
+++ b/drivers/gpu/drm/nouveau/include/nvif/head.h
@@ -3,6 +3,7 @@
 #define __NVIF_HEAD_H__
 #include <nvif/object.h>
 #include <nvif/event.h>
+#include <nvif/if0013.h>
 struct nvif_disp;
 
 struct nvif_head {
@@ -18,6 +19,8 @@ nvif_head_id(struct nvif_head *head)
 	return head->object.handle;
 }
 
+int nvif_head_armed(struct nvif_head *, struct nvif_head_armed_v0 *);
+
 int nvif_head_vblank_event_ctor(struct nvif_head *, const char *name, nvif_event_func, bool wait,
 				struct nvif_event *);
 #endif
diff --git a/drivers/gpu/drm/nouveau/include/nvif/if0013.h b/drivers/gpu/drm/nouveau/include/nvif/if0013.h
index 6756c7467ae4ef4de2e3fc658e16c96a2848e3db..dad552dc8946250566a775848988325154e7d68d 100644
--- a/drivers/gpu/drm/nouveau/include/nvif/if0013.h
+++ b/drivers/gpu/drm/nouveau/include/nvif/if0013.h
@@ -32,4 +32,38 @@ union nvif_head_scanoutpos_args {
 		__u16 hline;
 	} v0;
 };
+
+#define NVIF_HEAD_V0_ARMED 0x01
+
+union nvif_head_armed_args {
+	struct nvif_head_armed_v0 {
+		__u8  version;
+		__u8  vtaps;
+		__u8  htaps;
+		__u8  interlace;
+		/* Set when the class supplies viewport and tap readback
+		 * (GV100+). Ignore those fields when clear. This flag does not
+		 * report whether the head has a scaler.
+		 */
+		__u8  view;
+		__u8  nhsync;
+		__u8  nvsync;
+		/* GB20x tile ownership, zero where the class has none. */
+		__u8  tiles_mask;
+		__u64 hz;
+		__u16 htotal;
+		__u16 hsynce;
+		__u16 hblanke;
+		__u16 hblanks;
+		__u16 vtotal;
+		__u16 vsynce;
+		__u16 vblanke;
+		__u16 vblanks;
+		__u16 iW;
+		__u16 iH;
+		__u16 oW;
+		__u16 oH;
+		__u32 phywins[2];
+	} v0;
+};
 #endif
diff --git a/drivers/gpu/drm/nouveau/nvif/head.c b/drivers/gpu/drm/nouveau/nvif/head.c
index f00e01d232db436a865181d44bc87567d2debfd0..4119ea71966c5b277b89a81d59a11c2be05c74a0 100644
--- a/drivers/gpu/drm/nouveau/nvif/head.c
+++ b/drivers/gpu/drm/nouveau/nvif/head.c
@@ -26,6 +26,18 @@
 #include <nvif/class.h>
 #include <nvif/if0013.h>
 
+int
+nvif_head_armed(struct nvif_head *head, struct nvif_head_armed_v0 *args)
+{
+	int ret;
+
+	args->version = 0;
+
+	ret = nvif_mthd(&head->object, NVIF_HEAD_V0_ARMED, args, sizeof(*args));
+	NVIF_ERRON(ret && ret != -ENODEV, &head->object, "[ARMED] ret:%d", ret);
+	return ret;
+}
+
 int
 nvif_head_vblank_event_ctor(struct nvif_head *head, const char *name, nvif_event_func func,
 			    bool wait, struct nvif_event *event)
diff --git a/drivers/gpu/drm/nouveau/nvkm/engine/disp/uhead.c b/drivers/gpu/drm/nouveau/nvkm/engine/disp/uhead.c
index f072cec1604028e89365e5ac51a0545b24184f9a..cd1d2ab9801652564df65a7b354addc5faee29ae 100644
--- a/drivers/gpu/drm/nouveau/nvkm/engine/disp/uhead.c
+++ b/drivers/gpu/drm/nouveau/nvkm/engine/disp/uhead.c
@@ -71,6 +71,51 @@ nvkm_uhead_mthd_scanoutpos(struct nvkm_head *head, void *argv, u32 argc)
 	return 0;
 }
 
+static int
+nvkm_uhead_mthd_armed(struct nvkm_head *head, void *argv, u32 argc)
+{
+	union nvif_head_armed_args *args = argv;
+
+	if (argc != sizeof(args->v0) || args->v0.version != 0)
+		return -ENOSYS;
+
+	head->func->state(head, &head->arm);
+
+	/* VBIOS can leave nonzero raster totals on unused heads, so require a
+	 * pixel clock as well before exposing scanout state. On failure the
+	 * caller must ignore the remaining fields.
+	 */
+	if (!head->arm.htotal || !head->arm.vtotal || !head->arm.hz)
+		return -ENODEV;
+
+	args->v0.nhsync  = head->arm.or.nhsync;
+	args->v0.nvsync  = head->arm.or.nvsync;
+	args->v0.tiles_mask = head->arm.mtc.tiles;
+	args->v0.phywins[0] = head->arm.mtc.phywins[0];
+	args->v0.phywins[1] = head->arm.mtc.phywins[1];
+	args->v0.hz      = head->arm.hz;
+	args->v0.htotal  = head->arm.htotal;
+	args->v0.hsynce  = head->arm.hsynce;
+	args->v0.hblanke = head->arm.hblanke;
+	args->v0.hblanks = head->arm.hblanks;
+	args->v0.vtotal  = head->arm.vtotal;
+	args->v0.vsynce  = head->arm.vsynce;
+	args->v0.vblanke = head->arm.vblanke;
+	args->v0.vblanks = head->arm.vblanks;
+	/* Only classes with viewport readback set vtaps, making it a validity
+	 * marker even when the head is not scaling.
+	 */
+	args->v0.view    = head->arm.view.vtaps != 0;
+	args->v0.iW      = head->arm.view.iW;
+	args->v0.iH      = head->arm.view.iH;
+	args->v0.oW      = head->arm.view.oW;
+	args->v0.oH      = head->arm.view.oH;
+	args->v0.vtaps   = head->arm.view.vtaps;
+	args->v0.htaps   = head->arm.view.htaps;
+	args->v0.interlace = head->arm.interlace;
+	return 0;
+}
+
 static int
 nvkm_uhead_mthd(struct nvkm_object *object, u32 mthd, void *argv, u32 argc)
 {
@@ -78,6 +123,7 @@ nvkm_uhead_mthd(struct nvkm_object *object, u32 mthd, void *argv, u32 argc)
 
 	switch (mthd) {
 	case NVIF_HEAD_V0_SCANOUTPOS: return nvkm_uhead_mthd_scanoutpos(head, argv, argc);
+	case NVIF_HEAD_V0_ARMED: return nvkm_uhead_mthd_armed(head, argv, argc);
 	default:
 		return -EINVAL;
 	}
